RELEASE/scripts/autoscend/auto_quest_level_12.ash

```diff
@@ -1290,6 +1405,200 @@ boolean L12_themtharHills()
+		}
+	}
+	
+	//is it worth it to farm a chaos butterfly?
+	if(!get_property("chaosButterflyThrown").to_boolean() && item_amount($item[chaos butterfly]) == 0)
+	{
+		//softcore pull?
+		if(!in_hardcore() && adv_saved > adv_needed-15)
+		{
+			if(pullXWhenHaveY($item[chaos butterfly], 1, 0))
+			{
+				return true;
+			}
+		}
+		
+		//4 enemies in [The Castle in the Clouds in the Sky (Ground Floor)] mean there is a 25% chance to encounter the one we want.
+		//roughly estimate 4 turns per possibility giant encounter. at base drop this means ~20 adv needed.
+		if(is100FamiliarRun())
+		{
+			addToMaximize("200 item drop");
+		}
+		else
+		{
+			addToMaximize("200 item drop, switch Baby Gravy Fairy, switch Jumpsuited Hound Dog");
+		}
+		float expectedItemDropMulti = 1 + simValue("Item Drop")/100;
+		int adv_to_get_butterfly = ceil(20 / expectedItemDropMulti);
+		if(adv_saved > (adv_needed + adv_to_get_butterfly))
+		{
+			if(autoAdv(1, $location[The Castle in the Clouds in the Sky (Ground Floor)]))
+			{
+				return true;
+			}
+			else
+			{
+				auto_log_warning("For some reason failed to adventure in [The Castle in the Clouds in the Sky (Ground Floor)] for a [chaos butterfly]... skipping", "red");
+			}
+		}
+	}
+	
+	//if you just now acquired [chaos butterfly] and don't have anywhere to use it other than the barn or the war then use it in barn
+	//also contain code to account for being unable to use it for some reason
+	if(!get_property("chaosButterflyThrown").to_boolean() && item_amount($item[chaos butterfly]) > 0 && auto_my_path() != "Pocket Familiars")
+	{
+		if($location[McMillicancuddy's Barn].turns_spent > 0)
+		{
+			auto_log_warning("I seem to have spent turns in [McMillicancuddy's Barn] without using the [chaos butterfly] which I have. Something is not right...", "red");
+		}
+		else if(adv_saved > adv_needed-15)
+		{
+			auto_log_warning("Looks like I should use the [chaos butterfly] in [McMillicancuddy's Barn]", "blue");
+			if(autoAdv(1, $location[McMillicancuddy's Barn]))
+			{
+				return true;
+			}
+		}		
+	}
+	
+	//final check and then we can start doing the farm
+	if(adv_saved > adv_needed)
+	{
+		auto_log_info("Save McMillicancuddy's Farm from the Dooks", "blue");
+		set_property("choiceAdventure147", "3");	//open the pond
+		set_property("choiceAdventure148", "1");	//open the back 40
+		set_property("choiceAdventure149", "2");	//open the other back 40
+		
+		//there is no mafia tracking for stages of this sidequest
+		//because of free fights we can't rely on adventures spent count to see if a zone is clear
+		//instead we use the internal property auto_L12FarmStage to determine at what stage we are
+		
+		if(get_property("auto_L12FarmStage").to_int() == 0)
+		{
+			if(autoAdv(1, $location[McMillicancuddy's Barn]))
+			{
+				return true;
+			}
+			else
```

Comment:
Stylistic point:

```
if (b)
{
  foo()
  return true;
}
else
{
  bar();
}
```

can equivalently be replaced with

```
if (b)
{
  foo()
  return true;
}
bar();
```

This is just as clear, so long as the if-branch is short (although this can reduce clarity if it's a twenty-line branch).

Reply:
They are not equivalent. The second one is assuming success and will be caught in an infinite loop if foo(); fails
doing

return foo();
is viable because it will return true on success and false on failure. With false resulting in skipping to the next step on doTasks()
but in such a case we are unable to print a debug message informing the user of what broke

Reply:
I did not say `return foo()`, I agree it has the problems you described. An example of the change I asked for:
```
if(get_property("auto_L12FarmStage").to_int() == 1)
{
	if(autoAdv(1, $location[McMillicancuddy's Pond]))
	{
		return true;
	}
	set_property("auto_L12FarmStage", "2");
}
```

Reply:
Hopefully it's clear that this is equivalent to wrapping the `set_property` call in an else. If we `return true` every time we hit the if block, which we definitely do, the only way we run any code after the if block is if the if block didn't fire. So the else is redundant.

Reply:
About return foo();
yea, i was suggesting it as an alternative that works.

about the other code. yea, this one would work. the else is indeed redundant there since if it reaches that line it automatically is an else.

Reply:
change made 66835e3